File: apps/node/src/scenario/interpolate.test.ts
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import {
  interpolate,
  interpolateBody,
  interpolateHeaders,
  InterpolationError,
} from './interpolate.js';

describe('interpolate', () => {
  it('replaces {{var}} with the variable value', () => {
    assert.equal(interpolate('hello {{name}}', { name: 'world' }), 'hello world');
  });

  it('handles multiple variables in one string', () => {
    assert.equal(
      interpolate('{{greet}}, {{name}}!', { greet: 'hi', name: 'ada' }),
      'hi, ada!',
    );
  });

  it('leaves a plain string with no placeholders untouched', () => {
    assert.equal(interpolate('plain', {}), 'plain');
  });

  it('tolerates whitespace inside {{ name }}', () => {
    assert.equal(interpolate('{{ name }}', { name: 'x' }), 'x');
  });

  it('coerces non-string values to strings', () => {
    assert.equal(interpolate('id={{id}}', { id: 42 }), 'id=42');
    assert.equal(interpolate('ok={{flag}}', { flag: true }), 'ok=true');
  });

  it('supports dotted paths to nested objects', () => {
    assert.equal(
      interpolate('user={{user.id}}', { user: { id: 7 } }),
      'user=7',
    );
  });

  it('throws InterpolationError for missing variables', () => {
    assert.throws(() => interpolate('hi {{nope}}', {}), InterpolationError);
  });

  it('throws InterpolationError when a nested path misses', () => {
    assert.throws(
      () => interpolate('{{user.missing}}', { user: {} }),
      InterpolationError,
    );
  });

  it('throws InterpolationError for null/undefined values', () => {
    assert.throws(() => interpolate('{{x}}', { x: null }), InterpolationError);
    assert.throws(() => interpolate('{{x}}', { x: undefined }), InterpolationError);
  });
});

describe('interpolateHeaders', () => {
  it('returns undefined when headers is undefined', () => {
    assert.equal(interpolateHeaders(undefined, {}), undefined);
  });

  it('interpolates every header value', () => {
    const out = interpolateHeaders(
      { Authorization: 'Bearer {{token}}', 'X-User': '{{uid}}' },
      { token: 'abc', uid: 7 },
    );
    assert.deepEqual(out, { Authorization: 'Bearer abc', 'X-User': '7' });
  });
});

describe('interpolateBody', () => {
  it('passes undefined and null through', () => {
    assert.equal(interpolateBody(undefined, {}), undefined);
    assert.equal(interpolateBody(null, {}), null);
  });

  it('interpolates string leaves', () => {
    assert.equal(interpolateBody('hi {{name}}', { name: 'x' }), 'hi x');
  });

  it('recurses into objects and arrays', () => {
    const out = interpolateBody(
      { a: 'u={{u}}', b: [{ c: 'tok={{t}}' }, 'plain'] },
      { u: 'x', t: 'y' },
    );
    assert.deepEqual(out, { a: 'u=x', b: [{ c: 'tok=y' }, 'plain'] });
  });

  it('passes non-string primitives through unchanged', () => {
    assert.equal(interpolateBody(42, {}), 42);
    assert.equal(interpolateBody(true, {}), true);
  });

  it('does not mutate the input object', () => {
    const input = { a: 'v={{v}}' };
    interpolateBody(input, { v: 'x' });
    assert.equal(input.a, 'v={{v}}');
  });
});

File: apps/go-generator/internal/scenario/interpolate_test.go
package scenario

import (
	"reflect"
	"testing"
)

func TestInterpolate(t *testing.T) {
	type tc struct {
		name string
		in   string
		vars map[string]any
		want string
	}
	cases := []tc{
		{"single var", "hello {{name}}", map[string]any{"name": "world"}, "hello world"},
		{"multiple vars", "{{greet}}, {{name}}!", map[string]any{"greet": "hi", "name": "ada"}, "hi, ada!"},
		{"plain string", "plain", map[string]any{}, "plain"},
		{"whitespace tolerant", "{{ name }}", map[string]any{"name": "x"}, "x"},
		{"int coerce", "id={{id}}", map[string]any{"id": 42}, "id=42"},
		{"bool coerce", "ok={{flag}}", map[string]any{"flag": true}, "ok=true"},
		{"dotted nested", "user={{user.id}}", map[string]any{"user": map[string]any{"id": 7}}, "user=7"},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			got, err := Interpolate(c.in, c.vars)
			if err != nil {
				t.Fatalf("unexpected err: %v", err)
			}
			if got != c.want {
				t.Fatalf("got %q want %q", got, c.want)
			}
		})
	}

	t.Run("missing var errors", func(t *testing.T) {
		_, err := Interpolate("hi {{nope}}", map[string]any{})
		if !IsInterpolationError(err) {
			t.Fatalf("got %v, want InterpolationError", err)
		}
	})

	t.Run("nested missing errors", func(t *testing.T) {
		_, err := Interpolate("{{user.missing}}", map[string]any{"user": map[string]any{}})
		if !IsInterpolationError(err) {
			t.Fatalf("got %v, want InterpolationError", err)
		}
	})

	t.Run("nil/missing errors", func(t *testing.T) {
		_, err := Interpolate("{{x}}", map[string]any{"x": nil})
		if !IsInterpolationError(err) {
			t.Fatalf("got %v, want InterpolationError for nil", err)
		}
		_, err = Interpolate("{{x}}", map[string]any{})
		if !IsInterpolationError(err) {
			t.Fatalf("got %v, want InterpolationError for missing key", err)
		}
	})
}

func TestInterpolateHeaders(t *testing.T) {
	t.Run("nil passthrough", func(t *testing.T) {
		out, err := InterpolateHeaders(nil, map[string]any{})
		if err != nil || out != nil {
			t.Fatalf("got (%v,%v), want (nil,nil)", out, err)
		}
	})
	t.Run("interpolates every value", func(t *testing.T) {
		out, err := InterpolateHeaders(
			map[string]string{"Authorization": "Bearer {{token}}", "X-User": "{{uid}}"},
			map[string]any{"token": "abc", "uid": 7},
		)
		if err != nil {
			t.Fatalf("err: %v", err)
		}
		want := map[string]string{"Authorization": "Bearer abc", "X-User": "7"}
		if !reflect.DeepEqual(out, want) {
			t.Fatalf("got %v want %v", out, want)
		}
	})
}

func TestInterpolateBody(t *testing.T) {
	t.Run("nil passthrough", func(t *testing.T) {
		out, err := InterpolateBody(nil, map[string]any{})
		if err != nil || out != nil {
			t.Fatalf("got (%v,%v), want (nil,nil)", out, err)
		}
	})
	t.Run("string leaf", func(t *testing.T) {
		out, err := InterpolateBody("hi {{name}}", map[string]any{"name": "x"})
		if err != nil || out != "hi x" {
			t.Fatalf("got %v err=%v", out, err)
		}
	})
	t.Run("recurses into objects and arrays", func(t *testing.T) {
		input := map[string]any{
			"a": "u={{u}}",
			"b": []any{map[string]any{"c": "tok={{t}}"}, "plain"},
		}
		out, err := InterpolateBody(input, map[string]any{"u": "x", "t": "y"})
		if err != nil {
			t.Fatalf("err: %v", err)
		}
		want := map[string]any{
			"a": "u=x",
			"b": []any{map[string]any{"c": "tok=y"}, "plain"},
		}
		if !reflect.DeepEqual(out, want) {
			t.Fatalf("got %v want %v", out, want)
		}
	})
	t.Run("non-string primitives pass through", func(t *testing.T) {
		if v, _ := InterpolateBody(42, nil); v != 42 {
			t.Fatalf("int: got %v", v)
		}
		if v, _ := InterpolateBody(true, nil); v != true {
			t.Fatalf("bool: got %v", v)
		}
	})
	t.Run("does not mutate input", func(t *testing.T) {
		input := map[string]any{"a": "v={{v}}"}
		_, _ = InterpolateBody(input, map[string]any{"v": "x"})
		if input["a"] != "v={{v}}" {
			t.Fatalf("input mutated: %v", input)
		}
	})
}
